### PR TITLE
docs(cheatsheet/dart): generate from dart sources

### LINKED ISSUE #1906: [Dart] Cheatsheet: generate from Dart sources

### PR DESCRIPTION
**Pull request by chalin**
Fixes #1906.

### CODE REVIEW

**Comment by chalin** [comment]
@kwalrath : ready for review. I'm unsure who else.
@petebacondarwin : could you take a look at this new dgeni package? It is largely inspired by the TS cheatsheet package.
cc @wardbell

**Comment by chalin** [comment]
This PR depends on #1910.

tldr; Travis doc build is failing because I seem to have not fully turned off the Dart API and cheatsheet generation on Travis; but the cheatsheet _is_ the subject of this PR. So, I think it will be simpler to just say that this PR depends on #1910. The build works locally, but once #1910 is merged, it should build successfully on Travis too.

**Comment by kwalrath** [comment]
LGTM

File: tools/dart-api-builder/index.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    };
+  })
+
+  .config(function (parseTagsProcessor, getInjectables) {
+    const tagDefs = requireNgIoPkg('./tag-defs');
+    parseTagsProcessor.tagDefinitions =
+      parseTagsProcessor.tagDefinitions.concat(getInjectables(tagDefs));
+  })
+
+  .config(function (readTypeScriptModules, writeFilesProcessor, readFilesProcessor) {
```
Comment (petebacondarwin) [inline tools/dart-api-builder/index.js]: `readTypeScriptModules` not required?

File: tools/dart-api-builder/services/packageInfo.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+var fs = require('fs');
+var path = require('canonical-path');
+
+/**
+ * Load information about this project from the pubspec.yaml
+ * @return {Object} The package information
+ */
+module.exports = function packageInfo() {
+  const ngPath = '../angular-dart';
```
Comment (petebacondarwin) [inline tools/dart-api-builder/services/packageInfo.js]: This is hard coded here but is a constant in the index.js file. Perhaps it should be passed as some kind of property or dependency?

**Comment by petebacondarwin** [comment]
My feeling with this PR is that it is duplicating a lot of what is done to generate the cheatsheets for TS and JS. Perhaps there is some way of making that processing more generic to accommodate Dart as well?

File: tools/dart-api-builder/index.js (hunk 1 above)
Comment (chalin) [inline tools/dart-api-builder/index.js]: No, we only need to read read `'cheatsheet/*.md'` files as input; see line 49 below.

Comment (petebacondarwin) [inline tools/dart-api-builder/index.js]: Right, so you can remove this parameter

File: tools/dart-api-builder/services/packageInfo.js (hunk 2 above)
Comment (chalin) [inline tools/dart-api-builder/services/packageInfo.js]: I agree, that bothered me too, but it's hard coded in the same way on the TS side :)
See `tools/api-builder/angular.io-package/services/packageInfo.js` and that's good enough for me (for now).

**Comment by chalin** [comment]
> My feeling with this PR is that it is duplicating a lot of what is done to generate the cheatsheets for TS and JS.

Actually, I felt that this was a clean design that reused only the necessary pieces of the TS/JS doc processing chain.

 > Perhaps there is some way of making that processing more generic to accommodate Dart as well?

I thought of that too, but some of our other Dart work has been intrusive enough already. Have a separate package that reused TS-side components, seemed like the most TS friendly way to go.  One of the main problems is that cheatsheet generation on the TS side is interwoven with the API generation, when it doesn't need to be. If someone were to do a clean refactoring on the TS side, then indeed we could reuse that for Dart. Any volunteers :)

Aside from that, is this good to go?

File: tools/dart-api-builder/index.js (hunk 1 above)
Comment (chalin) [inline tools/dart-api-builder/index.js]: Oh, ... right (and `writeFilesProcessor` can go too). Done.

**Comment by petebacondarwin** [comment]
I think if it works then let's good with it

**Comment by kwalrath** [comment]
I merged because I heard (elsewhere) that Ward's good with this.

**Pull request merged**